components/brave_federated/learning_service.cc

```diff
@@ -0,0 +1,192 @@
+    VLOG(2) << "Learning rate applied from server: " << lr;
+  }
+
+  ModelSpec spec{
+      32,   // num_params
+      64,   // batch_size
+      lr,   // learning_rate
+      500,  // num_iterations
+      0.5   // threshold
+  };
+
+  if (static_cast<int>(task.GetParameters().at(0).size()) != spec.num_params &&
+      static_cast<int>(task.GetParameters().at(1).size()) != 1) {
+    VLOG(2) << "Task specifies a different model size than the client";
+    return;
+  }
+  VLOG(2) << "Task model and client model match!";
+
+  std::unique_ptr<Model> model = std::make_unique<Model>(spec);
+  model->SetWeights(task.GetParameters().at(0));
+  model->SetBias(task.GetParameters().at(1).at(0));
+  auto task_runner =
+      std::make_unique<FederatedTaskRunner>(task, std::move(model));
+
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      {base::TaskPriority::BEST_EFFORT,
+       base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN},
+      base::BindOnce(&LearningService::LoadDatasetAndRunTask,
+                     base::Unretained(this), std::move(task_runner)),
+      base::BindOnce(&LearningService::OnTaskResultComputed,
+                     weak_factory_.GetWeakPtr()));
+}
+
+TaskResult LearningService::LoadDatasetAndRunTask(
+    std::unique_ptr<FederatedTaskRunner> task_runner) {
+  auto synthetic_dataset = std::make_unique<SyntheticDataset>(500);
+  SyntheticDataset test_dataset = synthetic_dataset->SeparateTestData(50);
+
+  task_runner->SetTrainingData(synthetic_dataset->GetDataPoints());
+  task_runner->SetTestData(test_dataset.GetDataPoints());
+  VLOG(2) << "Model and data set. Task runner initialized.";
+
+  return task_runner->Run();
+}
+
+void LearningService::OnTaskResultComputed(TaskResult result) {
+  TaskResultList results;
+  results.push_back(result);
+  PostTaskResults(results);
+}
+
+void LearningService::PostTaskResults(TaskResultList results) {
+  for (const auto& result : results) {
+    communication_adapter_->PostTaskResult(
+        result, base::BindOnce(&LearningService::OnPostTaskResults,
+                               base::Unretained(this)));
+  }
+}
+
+void LearningService::OnPostTaskResults(TaskResultResponse response) {
+  post_results_backoff_entry_->InformOfRequest(response.IsSuccessful());
+
+  if (response.IsSuccessful()) {
+    VLOG(2) << "Task results posted successfully";
+  } else {
+    VLOG(2) << "Task results posting failed";
+  }
+
+  base::TimeDelta reconnect =
+      post_results_backoff_entry_->GetTimeUntilRelease();
+  reconnect_timer_ = std::make_unique<base::RetainingOneShotTimer>();
+  reconnect_timer_->Start(FROM_HERE, reconnect, this,
+                          &LearningService::GetTasks);
+  VLOG(2) << "Reconnecting in " << reconnect.InSeconds() << "s";
+}
+
+void LearningService::OnEligibilityChanged(bool is_eligible) {
+  if (is_eligible) {
+    StartParticipating();
+    VLOG(2) << "Eligibility changed, started participating.";
+  } else {
+    StopParticipating();
```

Comment:
Could it happen before `Init()` is called?

Reply:
O, I see that it couldn't be, we don't call `OnEligibilityChanged` externally (despite the fact it's a public method now).

Reply:
`OnEligibilityChanged` is called by the notifier (EligibilityService) so it should be public, right?

Reply:
Moved private functions in the private section 👍